### PR TITLE
Proposal: persistent volume selector

### PR DESCRIPTION
**Pull request by pmorie**
Partially replaces #17056.  Another proposal will follow dealing with dynamic provisioning on top of storage classes.

@kubernetes/sig-storage

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
It was not immediately obvious to me that a pv could only satisfy a single storage-class.  I would state that as an explicit use case.

**Comment by jsafrane** [comment]
1)
I understood the label selector as a way to have multiple labels specifying volume properties, like:
* I want a volume that has "fast" label and I don't care about anything else.
* I want a volume that has both "fast" and "cheap" labels.

Are these two use cases above valid? I think we can't cover them with single `storage-class` label.

2)
What is relation between `StorageClass` API resource and `storage-class` label? Will we deny any PV/PVC that references unknown storage class? Or is it just "optional" resource and we'll just match PV labels and PVC selectors?

**Comment by eparis** [comment]
I personally feel like `storageClassLabel` and any code that cares about it, is a bad pattern. I'm all for descriptive storage class objects (which could be extended to DP objects) but a single name just doesn't seem like 'the way' kube works....

**Comment by pmorie** [comment]
@eparis @jsafrane I talked about this with Eric yesterday and I can see the utility of being able to express an actual selector as an attribute of the storage class to make it more flexible and able to accommodate multi-label selectors that represent storage classes.  However, if we do that, we need to define exactly the mechanics of doing a storage-class based selection versus an ordinary label selection.

**Comment by eparis** [comment]
I say it should be 'actual label selection'.  Just like a nodeSelector.  saad at one point suggested a special kind of label selection to reduce the need for for a explosion of 'per label combination storage classes' but i think, for right now, that explosion is fine.  If you want the combination of 'fast/slow', 'expensive/cheap', and 'east/west' you need 8 storage class definitions to describe those combinations.

A PVC might only have fast/expensive and could match either east/west. Just like a label selector.

But a PVC with fast,expensive,east,blue could never match anything.

Maybe we re-think that last thing in the future as saad once suggested....

**Comment by pmorie** [comment]
@eparis @jsafrane @saad-ali @thockin An alternative to what I have proposed would be to have a label selector on the actual storage class itself, which would allow you to express some arbitrary selector involving however many labels you want.

In that case, the problem is how do you reference the storage class from the claim?  Based on the existing discussion it seemed as thought there was broad agreement that we need to be able to reference classes by name in the PVC.

**Comment by eparis** [comment]
I actively disagree that 'name' is a valuable identifier   :)

**Comment by pmorie** [comment]
@eparis As a user, how do I make a claim against a class, then?  Do I need to get the selector from the class' definition, and put it into my pod manually?

**Comment by pmorie** [comment]
I've rewritten the proposal to not dictate the labeling scheme, PTAL all.

**Comment by eparis** [comment]
To answer @jsafrane question, I think a single PV could satisfy multiple storage classes, if so labeled...

**Comment by eparis** [comment]
@k8s-bot e2e test this issue: #21484

File: docs/design/storage_classes.md [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+kind: StorageClass
+metadata:
+  name: my-storage-class
+  labels:
+    type: bar
+    color: foo
+description: A very special storage class
+```
+
+#### New API resource or ConfigMap?
```
Comment (bgrant0607) [inline docs/design/storage_classes.md]: Or ThirdPartyResource.

File: docs/design/storage_classes.md [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+For example, say that an administrator defines a `StorageClass` called `my-
+storage-class` described by the labels `type=bar` and `color=foo`:
+
+```yaml
+apiVersion: extensions/v1
+kind: StorageClass
+metadata:
+  name: my-storage-class
+  labels:
+    type: bar
```
Comment (bgrant0607) [inline docs/design/storage_classes.md]: Note that there could be multiple storage classes with identical labels.

File: docs/design/storage_classes.md [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+that class.
+
+A user must be able to view which storage classes are available.  Storage
+classes must be addressable by name to allow users to reference them simply.
+
+If a user specifies a storage class with a claim and one or more of the
+existing criteria (such as capacity), then a volume must satisfy all of those
+conditions to be bound to the claim.
+
+Label selectors are used through the Kubernetes API to describe relationships
```
Comment (bgrant0607) [inline docs/design/storage_classes.md]: +1 to Label Selector.

File: docs/design/storage_classes.md [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+things and imagine that AWS has two availability zones, `east` and `west`. Our
+administrators want to differentiate between volumes of the same type in these
+two zones, so they create a taxonomy of volumes like so:
+
+1.  `tin-west`
+2.  `tin-east`
+3.  `bronze-west`
+4.  `bronze-east`
+5.  `silver-west`
+6.  `silver-east`
```
Comment (bgrant0607) [inline docs/design/storage_classes.md]: Yeah, this is ugly.

**Comment by bgrant0607** [comment]
cc @davidopp @mml since similar issues have come up with disruption SLO classes

File: docs/design/storage_classes.md (hunk 4 above)
Comment (pmorie) [inline docs/design/storage_classes.md]: I agree -- I only presented it as an example of different individuals having different thought processes.  I can make a prettier example if you like; that will not be a huge problem.

File: docs/design/storage_classes.md [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+[kubernetes/17056](https://github.com/kubernetes/kubernetes/pull/17056), it
+was assumed that we would use `ConfigMap` to hold information about storage
+classes.  This would not be impossible to do, but there are several challenges
+due to the fact that `PersistentVolume` is a global resource and `ConfigMap`
+is namespaced.   We would need a virtual resource which appeared to be global
+but which was a view on a namespaced resource -- no such pattern is currently
+implemented anywhere in the system.  For that virtual resource, we would need
+to solve all the permission/policy issues in Kubernetes and downstream
+projects like OpenShift.  We would also likely need to define reserved key
+names for certain attributes of `StorageClass` within the `ConfigMap` `data`
```
Comment (bgrant0607) [inline docs/design/storage_classes.md]: There is increasing resistance to adding new APIs and I think StorageClass, in particular, is a hard sell given that it has almost no inherent meaning. 

How hard would be be, really, to choose a namespace and use ThirdPartyResource for such policy objects, at least for the foreseeable future, until the patterns and semantics are more clear? Or, if that's not palatable, TRP could be extended to support cluster-scoped resources.

Comment (pmorie) [inline docs/design/storage_classes.md]: @bgrant0607 I agree that it has almost no meaning in the context of this proposal -- however, it will have a very well defined meaning and widespread appeal (my subjective opinion here) once we add dynamic provisioning on top of it.

File: docs/design/storage_classes.md [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+
+```yaml
+apiVersion: extensions/v1
+kind: StorageClass
+metadata:
+  name: gold-west
+  labels:
+    storage-class: gold
+    region: west
+description: Provisioned IOPS SSD in the west region
```
Comment (bgrant0607) [inline docs/design/storage_classes.md]: This description could be an annotation, as we recommend for all other resources. 

Any properties that an admin didn't want users selecting on could also be expressed as distinct annotations.

This is a type that's 100% user-provided metadata.

File: docs/design/storage_classes.md (hunk 5 above)
Comment (bgrant0607) [inline docs/design/storage_classes.md]: I imagine we'll need other similar policy objects. I wouldn't want to create a content-free API for each one.

**Comment by bgrant0607** [comment]
TL;DR: Love the label selector idea. Don't like StorageClass.

File: docs/design/storage_classes.md (hunk 5 above)
Comment (pmorie) [inline docs/design/storage_classes.md]: @bgrant0607 I think it would be easier to look at this differently if we were at the beginning of a release cycle and it was anticipated that we would have both storage classes and dynamic provisioning in that release.  The real meat of storage classes is holding information about provisioning.

For the record, I am fine making the description an annotation.

However, I would like to avoid reasoning about the specifics of the API based solely upon this proposal (though given the timing that does seem unavoidable).

File: docs/design/storage_classes.md (hunk 6 above)
Comment (pmorie) [inline docs/design/storage_classes.md]: I'm fine with making it an annotation.  For other matters, expressed myself in https://github.com/kubernetes/kubernetes/pull/25413#discussion_r63778413

**Comment by pmorie** [comment]
@bgrant0607 @eparis @childsb @thockin @wattsteve @jsafrane I have pared this down to just the label selector for 1.3.  PTAL.

File: docs/proposals/volume-selectors.md [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+criteria important to them and select volumes based on these criteria.
+
+```go
+// PersistentVolumeClaimSpec describes the common attributes of storage devices
+// and allows a Source for provider-specific attributes
+type PersistentVolumeClaimSpec struct {
+    // Contains the types of access modes required
+    AccessModes []PersistentVolumeAccessMode `json:"accessModes,omitempty"`
+    // PersistentVolumeSelector is a selector which must be true for the claim to bind to a volume
+    LabelSelector *unversioned.LabelSelector `json:"labelSelector,omitempty"`
```
Comment (eparis) [inline docs/proposals/volume-selectors.md]: There seems to be some inconsistency between `LabelSelector` and `PersistentVolumeSelector` throughout.

Comment (pmorie) [inline docs/proposals/volume-selectors.md]: that's what i get for doing stuff at 3am :clock3:

**Comment by pmorie** [comment]
I will make a pass this morning to correct the inconsistencies around selector field name

**Comment by pmorie** [comment]
Field names are up to date.  PTAL @bgrant0607 @eparis @thockin

**Comment by jsafrane** [comment]
lgtm && matches implementation in #25917, but I'll let someone else label it (as API change)

**Comment by rata** [comment]
This seems great to me (as a user). Just curious (haven't thought this much), the selector will only have "matchLabels"? Or will it be more flexible like the replica set?

**Comment by pmorie** [comment]
@rata both `matchLabels` and `matchExpressions` are supported.

**Comment by eparis** [comment]
PR fails `hack/verify-generated-docs.sh`

**Comment by pmorie** [comment]
@eparis updated
